packages/agreement-process/src/services/agreementService.ts

```diff
@@ -1248,3 +1273,29 @@ async function addContractOnFirstActivation(
 
   return agreement;
 }
+
+async function getConsumerFromDelegations(
+  eserviceId: EServiceId,
+  delegationId: DelegationId | undefined,
+  organizationId: TenantId,
+  readModelService: ReadModelService
+): Promise<Tenant> {
+  const delegations =
+    await readModelService.getActiveConsumerDelegationsByEserviceId(eserviceId);
+
+  if (delegations.length > 0) {
+    if (!delegationId) {
+      throw missingDelegationId(organizationId, eserviceId);
+    }
+
+    const delegation = retrieveDelegation(delegations, delegationId);
+    assertRequesterIsDelegate(delegation.delegateId, organizationId);
+    return await retrieveTenant(delegation.delegatorId, readModelService);
+  }
+
+  if (delegationId) {
+    throw noActiveDelegations(eserviceId);
+  }
+
+  return await retrieveTenant(organizationId, readModelService);
+}
```

Comment:
I found this function quite confusing. It's called `getConsumerFromDelegation` but can also retrieve a Tenant using the `organizationId`.
Also, the `if` conditions are not so readable. I prefer first to make a check on the `delegationId` and then if exists try to get the Tenant from the delegation otherwise from the organizationId

Reply:
> I found this function quite confusing. It's called `getConsumerFromDelegation` but can also retrieve a Tenant using the `organizationId`. Also, the `if` conditions are not so readable. I prefer first to make a check on the `delegationId` and then if exists try to get the Tenant from the delegation otherwise from the organizationId

You're right, I should rewrite this logic, the problem is that we should prevent the delegator to access this route and check if there's a delegation even if a delegationId was not passed